### PR TITLE
[SYCL-MLIR] Change `sycl` dialect structure

### PR DESCRIPTION
**Pull request by victor-eds**
This splitting of types and dialect definition is needed in order to reuse these definitions in other files, e.g., in the attributes definition.

This also adds additional documentation targets for the dialect.

This commit applies the following changes to tablegen files:

- `SYCLBase.td`: Hosts base definitions for the dialect, i.e., the dialect itself and base interface declaration to be used by interfaces;
- `SYCLOpAttributes.td`: Hosts attributes definitions'
- `SYCLOps.td`: Hosts `sycl` operations and traits. This will include the rest, so it will be used in the `CMake` file generating the dialect;
- `SYCLTypes.td`: Hosts `sycl` type definitions;
- `SYCLMethodOpInterface.td`: Hosts `SYCLMethodOpInterface` definition. Each interface will be defined in its own file.

And renames the following headers:

- `SYCLOpAttributes.h` -> `SYCLAttributes.h`
- `SYCLOpsDialect.h` -> `SYCLDialect.h`
- `SYCLOpTraits.h` -> `SYCLTraits.h`
- `SYCLOpsTypes.h` -> `SYCLTypes.h`
- `SYCLMethodOpInterface.h` now contains the interface definition.

Similar changes are applied to `.cpp` files to mimic the headers structure.

### CODE REVIEW

File: mlir-sycl/include/mlir/Dialect/SYCL/IR/CMakeLists.txt [hunk 1]
```diff
@@ -1,14 +1,14 @@
 add_mlir_dialect(SYCLOps sycl)
-add_mlir_doc(SYCLOps SYCLOps Dialects/ -gen-op-doc -gen-op-interface-doc)
+add_mlir_doc(SYCLBase SYCLBase Dialects/ -gen-dialect-doc)
+add_mlir_doc(SYCLOps SYCLOps Dialects/ -gen-op-doc)
+add_mlir_doc(SYCLTypes SYCLTypes Dialects/ -gen-typedef-doc)
```
Comment (victor-eds) [inline mlir-sycl/include/mlir/Dialect/SYCL/IR/CMakeLists.txt:4]: Add documentation targets

**Review by sommerlukas — APPROVED** [review]
LGTM, I think it's good to move in this direction and adopt the layout of other dialects.

File: mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLTraits.h [hunk 2]
```diff
@@ -1,13 +1,13 @@
...
 //
 // MLIR-SYCL is under the Apache License v2.0 with LLVM Exceptions.
 // See https://llvm.org/LICENSE.txt for license information.
 // SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 //
 //===----------------------------------------------------------------------===//
 
-#ifndef MLIR_SYCL_OPS_TRAITS_H_
-#define MLIR_SYCL_OPS_TRAITS_H_
+#ifndef MLIR_SYCL_TRAITS_H_
```
[the remark this answers is not in the document]
Comment (etiotto) [reply mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLTraits.h:9]: Also "DIALECT" should be in the macro name ? As in:

`MLIR_DIALECT_SYCL_TRAITS_H`

File: mlir-sycl/lib/Dialect/IR/SYCLMethodOpInterface.cpp [hunk 3]
```diff
@@ -0,0 +1,13 @@
+//===--- SYCLMethodOpInterface.cpp ----------------------------------------===//
+//
+// MLIR-SYCL is under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/IR/BuiltinOps.h"
+
```
Comment (etiotto) [inline mlir-sycl/lib/Dialect/IR/SYCLMethodOpInterface.cpp:10]: [nit]: remove empty lines ?

File: mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLAttributes.td [hunk 4]
```diff
@@ -1,17 +1,17 @@
-//===-- SYCLOpsAttributes.td - SYCL dialect attribute definitions *tablegen===//
+//===-- SYCLAttributes.td - SYCL dialect attribute definitions ---*tablegen===//
```
Comment (whitneywhtsang) [inline mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLAttributes.td:1]:
```suggestion
//===-- SYCLAttributes.td - SYCL dialect attribute definitions *- tablegen -*=//
```

**Review by whitneywhtsang — APPROVED** [review]

File: mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLTraits.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (victor-eds) [reply mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLTraits.h:9]: I'm pushing a change standardizing the macro names. Great point.

**Comment by victor-eds** [comment]
@etiotto I also moved the `.cpp` to a `SYCL` directory, mimicking the headers structure.

**Comment by etiotto** [comment]
Good @victor-eds, there is a clang-format issue so the test didn't run. LGTM after test are run cleanly.

**Pull request merged**